src/main/java/games/strategy/engine/lobby/server/db/BannedUsernameController.java

```diff
@@ -1,34 +1,51 @@
 package games.strategy.engine.lobby.server.db;
 
+import static com.google.common.base.Preconditions.checkNotNull;
+
 import java.sql.Connection;
 import java.sql.PreparedStatement;
 import java.sql.ResultSet;
 import java.sql.SQLException;
 import java.sql.Timestamp;
 import java.time.Instant;
 
+import javax.annotation.Nullable;
+
+import games.strategy.engine.lobby.server.Moderator;
 import games.strategy.util.Tuple;
 
 /**
  * Utility class to create/read/delete banned usernames (there is no update).
  */
 public class BannedUsernameController extends TimedController implements BannedUsernameDao {
   @Override
-  public void addBannedUsername(final String username, final Instant banTill) {
-    if (banTill == null || banTill.isAfter(now())) {
+  public void addBannedUsername(final String username, final @Nullable Instant banTill, final Moderator moderator) {
+    checkNotNull(moderator);
 
-      try (Connection con = Database.getPostgresConnection();
-          PreparedStatement ps = con.prepareStatement("insert into banned_usernames (username, ban_till) values (?, ?)"
-              + " on conflict (username) do update set ban_till=excluded.ban_till")) {
-        ps.setString(1, username);
-        ps.setTimestamp(2, banTill != null ? Timestamp.from(banTill) : null);
-        ps.execute();
-        con.commit();
-      } catch (final SQLException sqle) {
-        throw new IllegalStateException("Error inserting banned username:" + username, sqle);
-      }
-    } else {
+    if (banTill != null && banTill.isBefore(now())) {
       removeBannedUsername(username);
+      return;
+    }
```

Comment:
I extracted this guard clause to avoid nesting on the primary code path.

Reply:
When you are ready to force all to re download and update.  AND all are on the same database!  I recommend that you wipe the slate clean.  What we do need before any of that tho.  Is a database that logs who was banned all their info and the mod that banned them and perhaps a reason.  Recorded automatically.

Reply:
@prastle 

> When you are ready to force all to re download and update. AND all are on the same database! I recommend that you wipe the slate clean.

Makes sense.

> What we do need before any of that tho. Is a database that logs who was banned all their info and the mod that banned them and perhaps a reason. Recorded automatically.

This PR does everything you mentioned except for including the ban reason.  As I mentioned above, I believe adding the ban reason will have to wait for the next incompatible release.

Reply:
It doesn't matter
until a forced download because
many refuse to re download and register
Also we need a list that records all anon that can be accessed If you are keeping anon login

Reply:
I would suggest making 5 generic reasons that can be clicked since having actual type is probably a pia.  Just a thought.  But either way until you force the users tyo use the new database this doesn't matter that much.  I do really appreciate the work as a mod tho its an awesome step towards the future
Thanks

Reply:
@prastle 

> It doesn't matter
until a forced download because
many refuse to re download and register

Sorry, I should have explained this PR better.  The audit changes don't require users to upgrade to a client that contains this PR.  It's all contained in the lobby code.  All that is required is a lobby upgrade, and you'll have your audit logs regardless of which version a user uses to connect. :smile:

> Also we need a list that records all anon that can be accessed If you are keeping anon login

is there an existing feature request for that?  If not, could you please open one on the forum?

Reply:
> I would suggest making 5 generic reasons that can be clicked since having actual type is probably a pia.

That's a good idea.  We'll probably leave the reason in the DB as free-form text, but include "canned reasons" in the UI.  Think a set of radio buttons with the last option being "type whatever you want."  I'm going to add this to the feature request on the forum so we don't forget.

Reply:
If this includes for example 3621 I am great with that.  But when they can still login with no records avail under twenty diff names or more it makes it very difficult for a mod to do anything without keeping their own database of every new name they see that causes trouble.  So how about you record that as well that can be accessed when needed by those with server access if your keeping anon login.  Please and thanks Pras.

Reply:
Also @ssoloff  to explain fully if I have come off harsh.  We have needed this for years.  The true problem tho is the anon login.  Thus why half the mod team quit.  hopefully we are on the way to a solution.

Reply:
>  to explain fully if I have come off harsh

@prastle Not at all.  I just have a hard time understanding a lot of these issues due to lack of experience in the lobby.  I appreciate you keeping them in the dev team's face. :smile:

Reply:
ty for your hard work.  I will try to come up with some generics but really it simple
1 bad name
2 politics
3 spamming
4 abusing another player
5 whatever
most are simply covered by being polite
then main pt is it needs to be recorded who when and why by what mod
if they decide not to post any info or reason

Reply:
as a final thought it would be great
if you could make it so they cant do the mute ban etc
without having clicked a reason :)